packages/Chem/src/rdkit-service/rdkit-service-worker-similarity.ts

```diff
@@ -10,50 +16,64 @@ export class RdKitServiceWorkerSimilarity extends RdKitServiceWorkerBase {
     super(module, webRoot);
   }
 
-  getFingerprints(fingerprintType: Fingerprint) {
-    if (this._rdKitMols === null)
+   /**
+   * Calculates fingerprints either on pre-created array of RDMols or creating RDMOls on the fly.
+   * 
+   * @param {Fingerprint} fingerprintType Type of Fingerprint
+   * @param {dict} string[] List of molecule strings to calculate fingerprints on. In case it is passed to function RDMols will be created on the fly
+   */
+
+  getFingerprints(fingerprintType: Fingerprint, dict?: string[]): IFingerprint[] {
+    if (this._rdKitMols === null && !dict)
       return [];
 
-    const fps: (Uint8Array | null)[] = [];
-    try {
-      switch (fingerprintType) {
-      case Fingerprint.Pattern:
-        for (let i = 0; i < this._rdKitMols.length; ++i) {
-          try {
-            if (!this._rdKitMols[i])
-              fps.push(null);
-            else
-              fps.push(this._rdKitMols[i]!.get_pattern_fp_as_uint8array());
-          } catch {
-            fps.push(null);
-            continue;
-          }
+    const fpLength = dict ? dict.length : this._rdKitMols!.length;
+    const fps = new Array<Uint8Array | null>(fpLength).fill(null);
+    for (let i = 0; i < fpLength; ++i) {
+      let mol: IMolContext | null = null;
+      if (dict) {
+        const item = dict[i];
+        if (item && item !== '') {
+          mol = getMolSafe(item, {}, this._rdKitModule);
         }
-        break;
-      case Fingerprint.Morgan:
-        for (let i = 0; i < this._rdKitMols.length; ++i) {
-          try{
-            if (!this._rdKitMols[i] || this._rdKitMols[i] && this._rdKitMols[i]?.is_qmol)
-              fps.push(null);
-            else
-              fps.push(this._rdKitMols[i]!.get_morgan_fp_as_uint8array(JSON.stringify({
-              radius: this._fpRadius,
-              nBits: this._fpLength,
-            })));
-          } catch (error) {
-            fps.push(null);
-            continue;
-          }
+      }
+      const rdMol = dict ? mol?.mol : this._rdKitMols![i];
+      const isQMol = dict ? mol?.isQMol : this._rdKitMols![i]?.is_qmol;
+      try {
+        switch (fingerprintType) {
+          case Fingerprint.Pattern:
+            try {
+              if (rdMol)
+                fps[i] = rdMol.get_pattern_fp_as_uint8array();
+            } catch {
+              //do nothing, fp is already null
+            }
+            break;
+          case Fingerprint.Morgan:
+            try {
+              if (rdMol && !isQMol)
+                fps[i] = rdMol.get_morgan_fp_as_uint8array(JSON.stringify({
```

Comment:
JSON.stringify(...) should be created outside of the loop and reused. It's very minor, but good style nonetheless.

Reply:
Done